pallets/messages/src/migration/v3.rs

```diff
@@ -0,0 +1,229 @@
+	storage::PrefixIterator,
+	traits::{Get, GetStorageVersion},
+	weights::WeightMeter,
+};
+use frame_system::pallet_prelude::BlockNumberFor;
+#[cfg(feature = "try-runtime")]
+use parity_scale_codec::Encode;
+
+const LOG_TARGET: &str = "pallet::messages::migration::v3";
+
+// Migration cursor: (storage index, cumulative records processed)
+#[allow(type_alias_bounds)]
+pub(crate) type MessagesCursor<T: Config> = (BlockNumberFor<T>, SchemaId, MessageIndex, u64);
+
+pub(crate) fn migrate_single_record<T: Config>(
+	iter: &mut PrefixIterator<(
+		(K1Type<T>, K2Type, K3Type),
+		v2::Message<T::MessagesMaxPayloadSizeBytes>,
+	)>,
+	cursor: &mut MessagesCursor<T>,
+) -> bool {
+	// If there's a next item in the iterator, perform the migration.
+	let messages_remain = if let Some(((block_number, schema_id, index), value)) = iter.next() {
+		// Migrate the inner value to the new structure
+		let new_value = Message {
+			schema_id, // schema_id is added to the struct
+			payload: value.payload,
+			provider_msa_id: value.provider_msa_id,
+			msa_id: value.msa_id,
+		};
+		// We can just insert here since the old and the new map share the same key-space.
+		MessagesV3::<T>::insert((block_number, schema_id, index), new_value);
+
+		cursor.3 += 1;
+		true
+	} else {
+		false
+	};
+
+	if !messages_remain || cursor.3 % <u64>::from(T::MigrateEmitEvery::get()) == 0 {
+		Pallet::<T>::deposit_event(Event::<T>::MessagesMigrated {
+			from_version: 2,
+			to_version: 3,
+			cumulative_total_migrated: cursor.3,
+		});
+	}
+
+	messages_remain
+}
+
+/// Migrates the items of the [`v2::MessagesV2`] map to [`crate::MessagesV3`]
+///
+/// The `step` function will be called once per block. It is very important that this function
+/// *never* panics and never uses more weight than it got in its meter. The migrations should also
+/// try to make maximal progress per step, so that the total time it takes to migrate stays low.
+pub struct MigrateV2ToV3<T: Config, W: weights::WeightInfo>(PhantomData<(T, W)>);
+impl<T: Config, W: weights::WeightInfo> SteppedMigration for MigrateV2ToV3<T, W> {
+	// Cursor type for migration. Not used for migration logic itself, other than
+	// for emitting/tracking migration progress/status.
+	// ((tuple of last message processed), cumulative message count)
+	type Cursor = MessagesCursor<T>;
+	// Without the explicit length here the construction of the ID would not be infallible.
+	type Identifier = MigrationId<31>;
+
+	/// The identifier of this migration. Which should be globally unique.
+	fn id() -> Self::Identifier {
+		MigrationId {
+			pallet_id: *b"pallet::messages::migration::v3",
+			version_from: 2,
+			version_to: 3,
+		}
+	}
+
+	/// The actual logic of the migration.
+	///
+	/// This function is called repeatedly until it returns `Ok(None)`, indicating that the
+	/// migration is complete. Ideally, the migration should be designed in such a way that each
+	/// step consumes as much weight as possible.
+	fn step(
```

Comment:
Note, this migration does NOT block the pallet's extrinsics from executing while in progress. This is because stored messages are indexed first by block number, and the pallet will only write going forward to the new `MessagesV3` storage; so there is ZERO risk of a key conflict between old records being migrated and new records written by the pallet.